File: src/Test/LazyEvaluationFFI.rs
use std::rc::Rc;

type Lazy<'a, A> = Rc<dyn Fn() -> A + 'a>;

fn force<A>(l: &Lazy<'_, A>) -> A {
    l()
}

fn defer<'a, A, F: Fn() -> A + 'a>(f: F) -> Lazy<'a, A> {
    Rc::new(f)
}

fn build_thunks<'a>(depth: i64, acc: Lazy<'a, i64>) -> Lazy<'a, i64> {
    if depth == 0 {
        return acc;
    }
    let prev = acc.clone();
    build_thunks(depth - 1, defer(move || force(&prev) + 1))
}

pub fn Test_LazyEvaluationFFI_runLazyEvaluationFFI(limit: i64) -> i64 {
    let mut acc = 0;
    for _ in 0..limit {
        let t = build_thunks(1000, defer(|| 0));
        acc += force(&t);
    }
    acc
}

File: src/Test/LazyEvaluationFFI.cc
#include "purescript.h"
#include <functional>
#include <memory>

namespace {
    template <class A> using LazyFn = std::function<A()>;
    template <class A> using Lazy = std::shared_ptr<const LazyFn<A>>;

    template <class A>
    Lazy<A> defer(std::function<A()> thunk) {
        return std::make_shared<LazyFn<A>>(std::move(thunk));
    }

    template <class A>
    A force(const Lazy<A>& value) {
        return (*value)();
    }

    Lazy<int> buildThunks(int depth, Lazy<int> acc) {
        while (depth > 0) {
            const Lazy<int> previous = acc;
            acc = defer<int>([previous]() { return force<int>(previous) + 1; });
            --depth;
        }
        return acc;
    }
}

FOREIGN_BEGIN(Test_LazyEvaluationFFI)
exports["runLazyEvaluationFFI"] = [](const boxed& in) -> boxed {
    const int repetitions = unbox<int>(in);
    int result = 0;
    for (int repetition = 0; repetition < repetitions; ++repetition) {
        result += force<int>(buildThunks(1000, defer<int>([]() { return 0; })));
    }
    return result;
};
FOREIGN_END
